src/lib/mongoBackend/mongoQueryContext.cpp

```diff
@@ -38,6 +38,268 @@
+        {
+          /* CRR without attributes (we need to keep searching in other CRR) */
+          perEntPa = crr->contextRegistration.providingApplication.get();
+          break; /* jx */
+        }
+        else
+        {
+          /* Is there a matching entity or the abcense of attributes? */
+          for (unsigned kx = 0; kx < crr->contextRegistration.contextRegistrationAttributeVector.size(); ++kx)
+          {
+            std::string regAttrName = crr->contextRegistration.contextRegistrationAttributeVector.get(kx)->name;
+            if (regAttrName == attrName)
+            {
+              /* We cannot "improve" this result keep searching in CRR vector, so we return */
+              perAttrPa = crr->contextRegistration.providingApplication.get();
+              return;
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+/* ****************************************************************************
+*
+* fillContextProviders -
+*
+* Looks in the elements of the CER vector passed as argument, searching for a suitable CPr in the CRR
+* vector passes as argument. If a suitable CPr is found, it is added to the CER (and the 'found' field
+* changed to true)
+*
+*/
+void fillContextProviders(ContextElementResponseVector& cerV, ContextRegistrationResponseVector& crrV)
+{
+  for (unsigned int ix = 0; ix < cerV.size(); ++ix)
+  {
+    ContextElementResponse* cer = cerV.get(ix);
+    for (unsigned int jx = 0; jx < cer->contextElement.contextAttributeVector.size(); ++jx)
+    {
+      ContextAttribute* ca = cer->contextElement.contextAttributeVector.get(jx);
+      if (!ca->found)
+      {        
+        /* Search for some CPr in crrV */
+        std::string perEntPa, perAttrPa;
+        searchCprForAttribute(cer->contextElement.entityId, ca->name, crrV, perEntPa, perAttrPa);
+
+        /* Looking results after crrV processing */
+        ca->providingApplication = perAttrPa == ""? perEntPa : perAttrPa;
+        ca->found = (ca->providingApplication != "");
+      }
+    }
+  }
+}
+
+/* ****************************************************************************
+*
+* addContextProviderEntity -
+*
+*/
+void addContextProviderEntity(ContextElementResponseVector& cerV, EntityId* enP, std::string pa)
+{
+  for (unsigned int ix = 0; ix < cerV.size(); ++ix)
+  {
+    if (cerV.get(ix)->contextElement.entityId.id == enP->id && cerV.get(ix)->contextElement.entityId.type == enP->type)
+    {
+      cerV.get(ix)->contextElement.providingApplicationList.push_back(pa);
+      return;    /* by construction, no more than one CER with the same entity information should exist in the CERV) */
+    }
+  }
+
+  /* Reached this point, it means that the cerV doesn't contain a proper CER, so we create it */
+  ContextElementResponse* cerP = new ContextElementResponse();
+  cerP->contextElement.entityId.id = enP->id;
+  cerP->contextElement.entityId.type = enP->type;
+  cerP->contextElement.entityId.isPattern = "false";
```

Comment:
align '='s

If malloc fails (no memory left), the broker will crash right here (cerP is NULL).
Malloc usually don't fail and I'm sure we have this problem is a hundred places.
No reason to add to it though :-)